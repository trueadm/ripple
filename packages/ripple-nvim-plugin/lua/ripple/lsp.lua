local M = {}

local SERVER_NAME = "ripple"
local LSP_PACKAGE = "ripple-language-server"
local EXACT_VERSION_PATTERN = "^%d+%.%d+%.%d+$"

local function plugin_package_json_path()
	local plugin_root = debug.getinfo(1, "S").source:sub(2)
	return vim.fs.find("package.json", { upward = true, path = vim.fs.dirname(plugin_root) })[1]
end

local function read_json(path)
	local stat = vim.loop.fs_stat(path)
	if not stat or stat.type ~= "file" then
		return nil
	end

	local ok, contents = pcall(vim.fn.readfile, path)
	if not ok or type(contents) ~= "table" then
		return nil
	end

	local json_string = table.concat(contents, "")

	local decoded
	local ok_decode, value = pcall(vim.json.decode, json_string)
	if ok_decode then
		decoded = value
	end

	if type(decoded) ~= "table" then
		return nil
	end

	return decoded
end

local function resolve_required_version()
	local package_json_path = plugin_package_json_path()

	local package_json = read_json(package_json_path)
	if not package_json then
		return nil, "unable to read plugin package.json"
	end

	local config = package_json.config or {}

	local spec = config["ripple-language-server"]

	if type(spec) ~= "string" or spec == "" then
		return nil, "missing config.rippleLanguageServerVersion field in package.json"
	end

	spec = spec:gsub("^%s+", ""):gsub("%s+$", "")

	if not spec:match(EXACT_VERSION_PATTERN) then
		return nil, ("unsupported version spec '%s'; please pin an exact semver"):format(spec)
	end

	return spec, nil
end

local function installed_version(install_dir)
	local pkg = read_json(install_dir .. "/node_modules/" .. LSP_PACKAGE .. "/package.json")
	if not pkg or type(pkg.version) ~= "string" then
		return nil
	end
	return pkg.version
end

local function ensure_server_binary()
	local required_version, err = resolve_required_version()
	if not required_version then
		vim.notify("[ripple] " .. err, vim.log.levels.ERROR)
		return nil
	end

	local install_dir = vim.fn.stdpath("data") .. "/" .. LSP_PACKAGE
	local bin = install_dir .. "/node_modules/.bin/" .. LSP_PACKAGE

	if vim.loop.os_uname().version:match("Windows") then
		bin = bin .. ".cmd"
	end

	local function version_matches()
		local version = installed_version(install_dir)
		return version == required_version
	end

	if vim.fn.executable(bin) == 1 and version_matches() then
		return bin
	end

	vim.fn.mkdir(install_dir, "p")
	vim.notify(("[ripple] Installing %s@%s..."):format(LSP_PACKAGE, required_version), vim.log.levels.INFO)

	local out = vim.fn.system({
		"npm",
		"install",
		("%s@%s"):format(LSP_PACKAGE, required_version),
		"--prefix",
		install_dir,
		"--no-audit",
		"--no-fund",
	})

	if vim.v.shell_error ~= 0 then
		vim.notify("[ripple] npm install failed:\n" .. out, vim.log.levels.ERROR)
		return nil
	end

	if not version_matches() then
		local found = installed_version(install_dir) or "unknown"
		vim.notify(
			("[ripple] Installed %s but required %s"):format(found, required_version),
			vim.log.levels.ERROR
		)
		return nil
	end

	return bin
end

function M.setup()
	local bin = ensure_server_binary()
	if not bin then
		return
	end

	local base_config = {
		cmd = { bin, "--stdio" },
		filetypes = { "ripple" },
		root_markers = { "package.json", "pnpm-workspace.yaml", ".git" },
	}

	vim.lsp.config(SERVER_NAME, base_config)
	vim.lsp.enable(SERVER_NAME)
end

return M
