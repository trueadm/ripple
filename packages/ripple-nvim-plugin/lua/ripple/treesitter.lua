local M = {}

function M.setup()
	local ok, parsers = pcall(require, "nvim-treesitter.parsers")
	if not ok then
		vim.schedule(function()
			vim.notify("[ripple] nvim-treesitter not found; skipping parser registration", vim.log.levels.WARN)
		end)
		return
	end

	local parser_config = parsers.get_parser_configs()
	local default_install = {
		url = "https://github.com/trueadm/ripple",
		files = {
			"src/parser.c",
			"src/scanner.c",
		},
		branch = "main",
		location = "packages/tree-sitter-ripple",
	}

	local config = parser_config.ripple or {}
	local install_info = config.install_info or {}
	config.install_info = vim.tbl_deep_extend("force", default_install, install_info)
	config.filetype = config.filetype or "ripple"
	config.used_by = config.used_by or { "ripple" }
	parser_config.ripple = config

	local ok_install, install = pcall(require, "nvim-treesitter.install")
	if ok_install and type(install.ensure_installed) == "function" then
		install.ensure_installed("ripple")
	else
		vim.schedule(function()
			vim.notify("[ripple] Install nvim-treesitter parser manually via :TSInstall ripple", vim.log.levels.INFO)
		end)
	end
end

return M
