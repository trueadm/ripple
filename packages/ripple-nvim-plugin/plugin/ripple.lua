if vim.filetype and vim.filetype.add then
	vim.filetype.add({
		extension = {
			ripple = "ripple",
		},
	})
end
