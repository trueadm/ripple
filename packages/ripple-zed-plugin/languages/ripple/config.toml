name = "Ripple"
grammar = "ripple"
path_suffixes = ["ripple"]
line_comments = ["//"]
block_comment = ["/*", "*/"]
tab_size = 2
hard_tabs = false

autoclose_before = ";:.,=}])\""

brackets = [
  { start = "{", end = "}", close = true, newline = true },
  { start = "[", end = "]", close = true, newline = true },
  { start = "(", end = ")", close = true, newline = true },
  { start = "<", end = ">", close = true, newline = false },
  { start = "\"", end = "\"", close = true, newline = false },
  { start = "'", end = "'", close = true, newline = false },
  { start = "`", end = "`", close = true, newline = false },
]
