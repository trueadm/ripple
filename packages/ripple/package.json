{
  "name": "ripple",
  "description": "Ripple is an elegant TypeScript UI framework",
  "license": "MIT",
  "author": "Dominic Gannaway",
  "version": "0.2.67",
  "type": "module",
  "module": "src/runtime/index.js",
  "main": "src/runtime/index.js",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/trueadm/ripple.git",
    "directory": "packages/ripple"
  },
  "bugs": {
    "url": "https://github.com/trueadm/ripple/issues"
  },
  "homepage": "https://ripplejs.com",
  "keywords": [
    "ripple",
    "UI",
    "framework"
  ],
  "exports": {
    ".": {
      "types": "./types/index.d.ts",
      "worker": "./src/runtime/index.js",
      "browser": "./src/runtime/index.js",
      "default": "./src/runtime/index.js"
    },
    "./package.json": "./package.json",
    "./compiler": {
      "types": "./types/index.d.ts",
      "require": "./compiler/index.js",
      "default": "./src/compiler/index.js"
    },
    "./validator": {
      "types": "./types/index.d.ts",
      "require": "./validator/index.js",
      "default": "./src/validator/index.js"
    },
    "./internal/client": {
      "default": "./src/runtime/internal/client/index.js"
    },
    "./jsx-runtime": {
      "types": "./src/jsx-runtime.d.ts",
      "import": "./src/jsx-runtime.js",
      "default": "./src/jsx-runtime.js"
    }
  },
  "imports": {
    "#client": "./src/runtime/internal/client/types.d.ts",
    "#client/constants": "./src/internal/client/constants.js"
  },
  "dependencies": {
    "@jridgewell/sourcemap-codec": "^1.5.5",
    "acorn": "^8.15.0",
    "acorn-typescript": "^1.4.13",
    "esrap": "^2.1.0",
    "is-reference": "^3.0.3",
    "magic-string": "^0.30.18",
    "muggle-string": "^0.4.1",
    "zimmerframe": "^1.1.2"
  },
  "devDependencies": {
    "@types/estree": "^1.0.8",
    "@types/node": "^24.3.0",
    "typescript": "catalog:"
  }
}
