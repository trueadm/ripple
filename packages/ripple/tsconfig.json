{
	"compilerOptions": {
		"module": "esnext",
		"lib": ["esnext", "dom", "dom.iterable"],
		"target": "esnext",
		"noEmit": true,
		"moduleResolution": "bundler",
		"resolveJsonModule": true,
		"noEmitOnError": true,
		"noErrorTruncation": true,
		"allowSyntheticDefaultImports": true,
		"verbatimModuleSyntax": true,
		"types": ["node"],
		"jsx": "preserve",
		"jsxImportSource": "ripple",
		"strict": true,
		"allowJs": true,
		"checkJs": true,
		"paths": {
			"ripple": ["./types/index.d.ts"],
		}
	},
	"include": [
		"./*.js",
		"./src/",
		"./tests/*/test.ripple",
	]
}
