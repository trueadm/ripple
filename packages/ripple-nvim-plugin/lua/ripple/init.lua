local M = {}

---Entrypoint invoked by plugin managers.
---@param opts? {treesitter?: table|boolean, lsp?: table|boolean}
function M.setup(opts)
	opts = opts or {treesitter = { ensure_installed = true }, lsp = {}}

	if opts.treesitter ~= false then
		require("ripple.treesitter").setup(opts.treesitter or {})
	end

	if opts.lsp ~= false then
		require("ripple.lsp").setup(opts.lsp or {})
	end
end

return M
