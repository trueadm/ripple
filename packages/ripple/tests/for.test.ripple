import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import { mount, flushSync, TrackedArray } from 'ripple';

describe('for statements', () => {
	let container;

	function render(component) {
		mount(component, {
			target: container
		});
	}

	beforeEach(() => {
		container = document.createElement('div');
		document.body.appendChild(container);
	});

	afterEach(() => {
		document.body.removeChild(container);
		container = null;
	});

	it('render a simple static array', () => {
		component App() {
			const items = ['Item 1', 'Item 2', 'Item 3'];

			for (const item of items) {
				<div class={item}>{item}</div>
			}
		}

		render(App);

		expect(container).toMatchSnapshot();
	});

	it('render a simple dynamic array', () => {
		component App() {
			const items = new TrackedArray('Item 1', 'Item 2', 'Item 3');

			for (const item of items) {
				<div class={item}>{item}</div>
			}

			<button onClick={() => items.push(`Item ${items.length + 1}`)}>{'Add Item'}</button>
		}

		render(App);
		expect(container).toMatchSnapshot();

		const button = container.querySelector('button');

		button.click();
		flushSync();

		expect(container).toMatchSnapshot();
	});

	it('correctly handles intermediate statements in for block', () => {
		component App() {
			const items = new TrackedArray(1, 2, 3);

			<div>
				for (const item of items) {
					<div>
						<div>{item}</div>
						const some_text = item;
						<div>{some_text}</div>
					</div>
				}
			</div>

			<button onClick={() => items.push(items.length + 1)}>{"Add Item"}</button>
		}

		render(App);

		expect(container).toMatchSnapshot();
		
		const button = container.querySelector('button');

		button.click();
		flushSync();

		expect(container).toMatchSnapshot();
	});

	it('correctly handle the index in a for...of loop', () => {
		component App() {
			const items = new TrackedArray('a', 'b', 'c');

			<div>
				for (let item of items; index i) {
					<div>{i + ' : ' + item}</div>
				}
			</div>

			<button onClick={() => items.push(String.fromCharCode(97 + items.length))}>{'Add Item'}</button>
			<button onClick={() => items.reverse()}>{'Reverse'}</button>
		}

		render(App);

		expect(container).toMatchSnapshot();

		const [button, button2] = container.querySelectorAll('button');

		button.click();
		flushSync();

		expect(container).toMatchSnapshot();

		button2.click();
		flushSync();

		expect(container).toMatchSnapshot();
	});
});
