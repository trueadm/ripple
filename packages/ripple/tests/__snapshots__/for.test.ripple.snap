// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`for statements > correctly handle the index in a for...of loop 1`] = `
<div>
  <div>
    <div>
      0 : a
    </div>
    <div>
      1 : b
    </div>
    <div>
      2 : c
    </div>
    
  </div>
  <button>
    Add Item
  </button>
  <button>
    Reverse
  </button>
  
</div>
`;

exports[`for statements > correctly handle the index in a for...of loop 2`] = `
<div>
  <div>
    <div>
      0 : a
    </div>
    <div>
      1 : b
    </div>
    <div>
      2 : c
    </div>
    <div>
      3 : d
    </div>
    
  </div>
  <button>
    Add Item
  </button>
  <button>
    Reverse
  </button>
  
</div>
`;

exports[`for statements > correctly handle the index in a for...of loop 3`] = `
<div>
  <div>
    <div>
      0 : d
    </div>
    <div>
      1 : c
    </div>
    <div>
      2 : b
    </div>
    <div>
      3 : a
    </div>
    
  </div>
  <button>
    Add Item
  </button>
  <button>
    Reverse
  </button>
  
</div>
`;

exports[`for statements > correctly handles intermediate statements in for block 1`] = `
<div>
  <div>
    <div>
      <div>
        1
      </div>
      <div>
        1
      </div>
    </div>
    <div>
      <div>
        2
      </div>
      <div>
        2
      </div>
    </div>
    <div>
      <div>
        3
      </div>
      <div>
        3
      </div>
    </div>
    
  </div>
  <button>
    Add Item
  </button>
  
</div>
`;

exports[`for statements > correctly handles intermediate statements in for block 2`] = `
<div>
  <div>
    <div>
      <div>
        1
      </div>
      <div>
        1
      </div>
    </div>
    <div>
      <div>
        2
      </div>
      <div>
        2
      </div>
    </div>
    <div>
      <div>
        3
      </div>
      <div>
        3
      </div>
    </div>
    <div>
      <div>
        4
      </div>
      <div>
        4
      </div>
    </div>
    
  </div>
  <button>
    Add Item
  </button>
  
</div>
`;

exports[`for statements > render a simple dynamic array 1`] = `
<div>
  <!---->
  <div
    class="Item 1"
  >
    Item 1
  </div>
  <div
    class="Item 2"
  >
    Item 2
  </div>
  <div
    class="Item 3"
  >
    Item 3
  </div>
  <!---->
  <button>
    Add Item
  </button>
  
</div>
`;

exports[`for statements > render a simple dynamic array 2`] = `
<div>
  <!---->
  <div
    class="Item 1"
  >
    Item 1
  </div>
  <div
    class="Item 2"
  >
    Item 2
  </div>
  <div
    class="Item 3"
  >
    Item 3
  </div>
  <div
    class="Item 4"
  >
    Item 4
  </div>
  <!---->
  <button>
    Add Item
  </button>
  
</div>
`;

exports[`for statements > render a simple static array 1`] = `
<div>
  <!---->
  <div
    class="Item 1"
  >
    Item 1
  </div>
  <div
    class="Item 2"
  >
    Item 2
  </div>
  <div
    class="Item 3"
  >
    Item 3
  </div>
  <!---->
  
</div>
`;
